distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Upload_Table.enso

```diff
@@ -229,20 +225,58 @@ common_update_table source_table connection table_name update_action key_columns
+            Update_Action.Update_Or_Insert -> throw_not_implemented
+            Update_Action.Align_Records -> throw_not_implemented
+        upload_status.if_not_error target_table
+
+## PRIVATE
+check_update_arguments source_table target_table key_columns update_action error_on_missing_columns on_problems ~action =
     source_columns = Set.from_vector source_table.column_names
     target_columns = Set.from_vector target_table.column_names
     extra_columns = source_columns.difference target_columns
-    if extra_columns.not_empty then Error.throw (Unmatched_Columns.Error extra_columns) else
+    if extra_columns.not_empty then Error.throw (Unmatched_Columns.Error extra_columns.to_vector) else
         missing_columns = target_columns.difference source_columns
-        if missing_columns.not_empty && error_on_missing_columns then Error.throw (Missing_Input_Columns.Error missing_columns "the source table") else
-            # TODO [RW] to be finished in follow up PR
-            _ = [update_action, key_columns]
-            Error.throw (Illegal_State.Error "TODO: Not implemented yet.")
+        if missing_columns.not_empty && error_on_missing_columns then Error.throw (Missing_Input_Columns.Error missing_columns.to_vector "the source table") else
+            key_set = Set.from_vector key_columns
+            missing_source_key_columns = key_set.difference source_columns
+            missing_target_key_columns = key_set.difference target_columns
+            # TODO Shall we add some 'location' to `Missing_Input_Columns` to distinguish when a missing column comes from table structure VS `key_columns`?
+            if missing_source_key_columns.not_empty then Error.throw (Missing_Input_Columns.Error missing_source_key_columns.to_vector "the source table") else
+                if missing_target_key_columns.not_empty then Error.throw (Missing_Input_Columns.Error missing_target_key_columns.to_vector "the target table") else
+                    if (update_action != Update_Action.Insert) && key_columns.is_empty then Error.throw (Illegal_Argument.Error "For the `update_action = "+update_action.to_text+"`, the `key_columns` must be specified to define how to match the records.") else
+                        # Verify type matching
+                        problems = source_table.columns.flat_map source_column->
+                            # The column must exist because it was verified earlier.
+                            target_column = target_table.get source_column.name
+                            source_type = source_column.value_type
+                            target_type = target_column.value_type
+                            if source_type == target_type then [] else
+                                if source_type.can_be_widened_to target_type then [Inexact_Type_Coercion.Warning source_type target_type] else
+                                    Error.throw (Column_Type_Mismatch.Error source_column.name target_type source_type)
```

Comment:
could we make this a local function - the nesting level is very high by this point!